tendermint/src/evidence.rs

```diff
@@ -2,50 +2,32 @@
 
 use std::slice;
 use {
-    crate::serializers,
-    serde::{de::Error as _, Deserialize, Deserializer, Serialize, Serializer},
-    subtle_encoding::base64,
+    crate::{serializers, PublicKey, Vote},
+    serde::{Deserialize, Serialize},
 };
 
 /// Evidence of malfeasance by validators (i.e. signing conflicting votes).
 /// encoded using an Amino prefix. There is currently only a single type of
 /// evidence: `DuplicateVoteEvidence`.
 ///
 /// <https://github.com/tendermint/tendermint/blob/master/docs/spec/blockchain/blockchain.md#evidence>
-#[derive(Clone, Debug)]
-pub struct Evidence(Vec<u8>);
-
-impl Evidence {
-    /// Create a new raw evidence value from a byte vector
-    pub fn new<V>(into_vec: V) -> Evidence
-    where
-        V: Into<Vec<u8>>,
-    {
-        // TODO(tarcieri): parse/validate evidence contents from amino messages
-        Evidence(into_vec.into())
-    }
-
-    /// Serialize this evidence as an Amino message bytestring
-    pub fn to_amino_bytes(&self) -> Vec<u8> {
-        self.0.clone()
-    }
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(tag = "type", content = "value")]
+pub enum Evidence {
+    /// Duplicate vote evidence
+    #[serde(rename = "tendermint/DuplicateVoteEvidence")]
+    DuplicateVote(DuplicateVoteEvidence),
 }
 
-impl<'de> Deserialize<'de> for Evidence {
-    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
-        let bytes = base64::decode(String::deserialize(deserializer)?.as_bytes())
-            .map_err(|e| D::Error::custom(format!("{}", e)))?;
-
-        Ok(Evidence::new(bytes))
-    }
-}
-
-impl Serialize for Evidence {
-    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
-        String::from_utf8(base64::encode(self.to_amino_bytes()))
-            .unwrap()
-            .serialize(serializer)
-    }
+/// Duplicate vote evidence
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct DuplicateVoteEvidence {
+    #[serde(rename = "PubKey")]
+    pub_key: PublicKey,
```

Comment:
I just noted that in the go versions there is no pubkey field (anymore?). We need to remove this for compatibility. Tracking here: https://github.com/informalsystems/tendermint-rs/issues/300